src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -0,0 +1,506 @@
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Order\CommandHandler;
+
+use Address;
+use Attribute;
+use Carrier;
+use Cart;
+use CartRule;
+use Combination;
+use Configuration;
+use Context;
+use Customer;
+use Hook;
+use Order;
+use OrderCarrier;
+use OrderCartRule;
+use OrderDetail;
+use OrderInvoice;
+use PrestaShop\PrestaShop\Adapter\Order\AbstractOrderHandler;
+use PrestaShop\PrestaShop\Core\Domain\Order\Command\AddProductToOrderCommand;
+use PrestaShop\PrestaShop\Core\Domain\Order\CommandHandler\AddProductToOrderHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use Product;
+use Shop;
+use SpecificPrice;
+use StockAvailable;
+use Symfony\Component\Translation\TranslatorInterface;
+use Tools;
+use Validate;
+
+/**
+ * Handles adding product to an existing order using legacy object model classes.
+ *
+ * @internal
+ */
+final class AddProductToOrderHandler extends AbstractOrderHandler implements AddProductToOrderHandlerInterface
+{
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @param TranslatorInterface $translator
+     */
+    public function __construct(TranslatorInterface $translator)
+    {
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(AddProductToOrderCommand $command)
+    {
+        $order = $this->getOrderObject($command->getOrderId());
+
+        $this->assertOrderWasNotShipped($order);
+
+        $product = $this->getProductObject($command->getProductId(), (int) $order->id_lang);
+        $combination = $this->getCombination($command->getCombinationId());
+
+        $cart = $this->createNewCart($order);
+
+        $specificPrice = $this->createSpecificPriceIfNeeded(
+            $command,
+            $order,
+            $cart,
+            $product,
+            $combination
+        );
+
+        $this->addProductToCart($cart, $product, $combination, $command->getProductQuantity());
+
+        $invoice = $this->createNewOrEditExistingInvoice(
+            $command,
+            $order,
+            $cart
+        );
+
+        $totalMethod = $command->getOrderInvoiceId() ? Cart::BOTH_WITHOUT_SHIPPING : Cart::BOTH;
+
+        // Create Order detail information
+        $orderDetail = new OrderDetail();
+        $orderDetail->createList(
+            $order,
+            $cart,
+            $order->getCurrentOrderState(),
+            $cart->getProducts(),
+            $command->getOrderInvoiceId() ?: 0
+        );
+
+        // update totals amount of order
+        $order->total_products += (float) $cart->getOrderTotal(false, Cart::ONLY_PRODUCTS);
```

Comment:
All of this was being done in the legacy controller 😮 ?

Reply:
Yup. :1st_place_medal: